In [14]:
import pandas as pd
import re

# Ruta local
ruta = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Inputs\noticias_politicas_procesadas.csv"

# Cargar dataset
df = pd.read_csv(ruta)

# Función para extraer fecha del inicio del contenido (formato: ddmmaaaa)
def extraer_fecha(texto):
    match = re.search(r"(\d{2})(\d{2})(\d{4})", texto)
    if match:
        return pd.to_datetime(f"{match.group(3)}-{match.group(2)}-{match.group(1)}")
    return pd.NaT

# Aplicar extracción de fecha
df["fecha"] = df["Contenido"].apply(extraer_fecha)

# Eliminar filas sin fecha
df = df.dropna(subset=["fecha"])

# Revisar
print(df[["Diario", "Título", "fecha"]].head())
df

           Diario                                             Título  \
0  Diario Gestión  Banco de la Nación le extiende el brazo a Petr...   
1  Diario Gestión  Banco de la Nación le extiende el brazo a Petr...   
2  Diario Gestión  Machu Picchu en emergencia?: los cambios que p...   
3  Diario Gestión  Wallenberg Investments AB adelanta sus planes ...   
4  Diario Gestión  Protesta de mineros informales en Arequipa: un...   

       fecha  
0 2025-07-11  
1 2025-07-11  
2 2025-07-11  
3 2025-07-11  
4 2025-07-11  


Diario                                             Título  \
0    Diario Gestión  Banco de la Nación le extiende el brazo a Petr...   
1    Diario Gestión  Banco de la Nación le extiende el brazo a Petr...   
2    Diario Gestión  Machu Picchu en emergencia?: los cambios que p...   
3    Diario Gestión  Wallenberg Investments AB adelanta sus planes ...   
4    Diario Gestión  Protesta de mineros informales en Arequipa: un...   
..              ...                                                ...   
327   Diario Correo  PCM: Hay ministerios que piden impugnar resolu...   
328   Diario Correo  Gobierno quiere modelo boliviano en sector hid...   
329   Diario Correo  Jorge Zapata, presidente de la Confiep: Es evi...   
330   Diario Correo  Convenios Gobierno a Gobierno deben incluir me...   
331   Diario Correo  Indecopi multa a Rappi con S 243 mil por inclu...   

                                                Enlace  \
0    https://gestion.pe/economia/banco-de-la-nacion...   
1    https://gestion.pe/economia/banco-de-la-nacion...   
2    https://gestion.pe/economia/machu-picchu-en-em...   
3    https://gestion.pe/economia/empresas/entrevist...   
4    https://gestion.pe/economia/pnp-libero-tramo-d...   
..                                                 ...   
327  https://diariocorreo.pe/economia/pcm-hay-minis...   
328  https://diariocorreo.pe/economia/gobierno-quie...   
329  https://diariocorreo.pe/economia/jorge-zapata-...   
330  https://diariocorreo.pe/economia/convenios-gob...   
331  https://diariocorreo.pe/economia/indecopi-mult...   

                                             Contenido  \
0    11072025 21H14 En los últimos días, el Gobiern...   
1    11072025 21H14 En los últimos días, el Gobiern...   
2    Fernando Cuadros Concha 11072025 20H25 La banc...   
3    Carlos Rosales Salas 11072025 17H31 El empresa...   
4    Redacción Gestión 11072025 17H18 La Policía Na...   
..                                                 ...   
327  Clorinda Flores Actualizado el 01042025, 07:00...   
328  Clorinda Flores Actualizado el 31032025, 02:54...   
329  Clorinda Flores Actualizado el 30032025, 09:33...   
330  Clorinda Flores Actualizado el 29032025, 08:00...   
331  Andrea Pereyra Actualizado el 28032025, 03:54 ...   

                                        titular_limpio  \
0    Banco de la nación le extiende el brazo a petr...   
1    Banco de la nación le extiende el brazo a petr...   
2    Machu picchu en emergencia los cambios que pla...   
3    Wallenberg investments ab adelanta sus planes ...   
4    Protesta de mineros informales en arequipa una...   
..                                                 ...   
327  Pcm hay ministerios que piden impugnar resoluc...   
328  Gobierno quiere modelo boliviano en sector hid...   
329  Jorge zapata presidente de la confiep es evide...   
330  Convenios gobierno a gobierno deben incluir me...   
331  Indecopi multa a rappi con mil por incluir clá...   

                                        resumen_limpio  \
0    En los últimos días el gobierno de perú ha hab...   
1    En los últimos días el gobierno de perú ha hab...   
2    Fernando cuadros concha la bancada de acción p...   
3    Carlos rosales salas el empresario sueco marcu...   
4    Redacción gestión la policía nacional del perú...   
..                                                 ...   
327  Clorinda flores actualizado el a según el titu...   
328  Clorinda flores actualizado el el ex viceminis...   
329  Clorinda flores actualizado el a para el presi...   
330  Clorinda flores actualizado el a el director e...   
331  Andrea pereyra actualizado el la sala especial...   

                                        titular_tokens  \
0    ['Banco', 'nación', 'extiende', 'brazo', 'petr...   
1    ['Banco', 'nación', 'extiende', 'brazo', 'petr...   
2    ['Machu', 'picchu', 'emergencia', 'cambios', '...   
3    ['Wallenberg', 'investments', 'ab', 'adelanta'...   
4    ['Protesta', 'mineros', 'informales', 'ar


Chatbot basado en reglas o plantillas

In [2]:
# Extraer fecha desde la columna Contenido
def extraer_fecha(texto):
    match = re.search(r"(\d{2})(\d{2})(\d{4})", str(texto))
    if match:
        return pd.to_datetime(f"{match.group(3)}-{match.group(2)}-{match.group(1)}")
    return pd.NaT

df["fecha"] = df["Contenido"].apply(extraer_fecha)
df = df.dropna(subset=["fecha"])

# === 2. Funciones base ===

def ultimas_noticias_por_diario(df, n=3):
    noticias = (
        df.sort_values("fecha", ascending=False)
          .groupby("Diario")
          .head(n)
          .sort_values(["Diario", "fecha"], ascending=[True, False])
    )
    return noticias[["Diario", "Título", "fecha", "Enlace"]]

def buscar_noticias_por_tema(df, tema, n=5):
    filtro = df["resumen_limpio"].str.contains(tema.lower(), case=False, na=False)
    resultados = df[filtro].sort_values("fecha", ascending=False).head(n)
    return resultados[["Diario", "Título", "fecha", "Enlace"]]

# === 3. Chatbot de consola ===

def responder_pregunta(pregunta, df):
    pregunta = pregunta.lower()
    
    if "nuevo" in pregunta or "últimas" in pregunta:
        noticias = ultimas_noticias_por_diario(df)
        respuesta = "📰 Últimas noticias políticas por diario:\n\n"
        for _, row in noticias.iterrows():
            respuesta += f"- ({row['Diario']}) {row['Título']} ({row['fecha'].date()})\n  {row['Enlace']}\n\n"
        return respuesta
    
    elif "dicen" in pregunta or "sobre" in pregunta:
        tema = pregunta.split("sobre")[-1].strip()
        noticias = buscar_noticias_por_tema(df, tema)
        if noticias.empty:
            return f"No encontré noticias recientes sobre “{tema}”."
        respuesta = f"🗞 Noticias recientes sobre “{tema}”:\n\n"
        for _, row in noticias.iterrows():
            respuesta += f"- ({row['Diario']}) {row['Título']} ({row['fecha'].date()})\n  {row['Enlace']}\n\n"
        return respuesta
    
    else:
        return (
            "🤖 No entendí tu consulta. Puedes probar con:\n"
            "- ¿Qué hay de nuevo en política?\n"
            "- Dime las últimas noticias políticas\n"
            "- ¿Qué dicen los periódicos sobre Dina Boluarte?"
        )

# === 4. Loop de prueba en consola ===

if __name__ == "__main__":
    print("🗣 Chatbot de noticias políticas. Escribe 'salir' para terminar.\n")
    while True:
        user_input = input("Tú: ")
        if user_input.lower() in ["salir", "exit", "quit"]:
            print("👋 Hasta luego.")
            break
        respuesta = responder_pregunta(user_input, df)
        print("\n" + respuesta + "\n")

🗣 Chatbot de noticias políticas. Escribe 'salir' para terminar.


📰 Últimas noticias políticas por diario:

- (Diario Correo) Un arancel de 50 al cobre hará caer exportación de Perú: Trump anunció que sería desde el 1 de agosto (2025-07-11)
  https://diariocorreo.pe/economia/un-arancel-de-50-al-cobre-hara-caer-exportacion-de-peru-trump-anuncio-que-seria-desde-el-1-de-agosto-noticia/

- (Diario Correo) Trabajadores deben recibir su gratificación de Fiestas Patrias hasta el 15 de julio (2025-07-11)
  https://diariocorreo.pe/economia/trabajadores-deben-recibir-su-gratificacion-de-fiestas-patrias-hasta-el-15-de-julio-peru-noticia/

- (Diario Correo) ADEX condena bloqueos y exige al Gobierno restablecer el orden (2025-07-11)
  https://diariocorreo.pe/economia/adex-condena-bloqueos-y-exige-al-gobierno-restablecer-el-orden-noticia/

- (Diario El Comercio) Adex pide al Gobierno restituir el orden tras bloqueos de carreteras en diversas regiones del país (2025-07-11)
  https://elcomercio.pe/eco

Usamos FLAN

In [3]:
pip install transformers sentencepiece accelerate


   ---------------------------------------- 0.0/991.5 kB ? eta -:--:--
   -- ------------------------------------- 61.4/991.5 kB 1.7 MB/s eta 0:00:01
   -- ------------------------------------- 71.7/991.5 kB 1.3 MB/s eta 0:00:01
   -- ------------------------------------- 71.7/991.5 kB 1.3 MB/s eta 0:00:01
   ----- -------------------------------- 153.6/991.5 kB 833.5 kB/s eta 0:00:02
   ------- ------------------------------ 194.6/991.5 kB 985.7 kB/s eta 0:00:01
   ------- ------------------------------ 194.6/991.5 kB 985.7 kB/s eta 0:00:01
   ------- ------------------------------ 194.6/991.5 kB 985.7 kB/s eta 0:00:01
   ------- ------------------------------ 194.6/991.5 kB 985.7 kB/s eta 0:00:01
   ------- ------------------------------ 194.6/991.5 kB 985.7 kB/s eta 0:00:01
   ---------- --------------------------- 286.7/991.5 kB 632.7 kB/s eta 0:00:02
   ------------ ------------------------- 337.9/991.5 kB 677.0 kB/s eta 0:00:01
   ------------------ ------------------- 481.3/991.


[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\HP\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [ ]:
##¿Dónde se conecta con FLAN-T5?

import pandas as pd
import re

# Cargar CSV
ruta = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Inputs\noticias_politicas_procesadas.csv"
df = pd.read_csv(ruta)

# Limpiar nombres de columnas
df.columns = df.columns.str.strip()

# Asegurar columna 'resumen_limpio'
if 'resumen_limpio' not in df.columns:
    print("⚠️ 'resumen_limpio' no encontrado. Usando 'Contenido' como respaldo.")
    df['resumen_limpio'] = df['Contenido']

# Extraer fecha desde Contenido
def extraer_fecha(texto):
    match = re.search(r"(\d{2})(\d{2})(\d{4})", str(texto))
    if match:
        return pd.to_datetime(f"{match.group(3)}-{match.group(2)}-{match.group(1)}")
    return pd.NaT

df["fecha"] = df["Contenido"].apply(extraer_fecha)
df = df.dropna(subset=["fecha"])


In [ ]:
# 1. Librerías
import pandas as pd
import re
import torch
from transformers import T5Tokenizer, T5ForConditionalGeneration

# 2. Cargar CSV y preparar el DataFrame
ruta = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Inputs\noticias_politicas_procesadas.csv"
df = pd.read_csv(ruta)
df.columns = df.columns.str.strip()

# Verificación de columna 'resumen_limpio'
if 'resumen_limpio' not in df.columns:
    print("⚠️ 'resumen_limpio' no encontrado. Usando 'Contenido' como respaldo.")
    df['resumen_limpio'] = df['Contenido']

# Extraer fecha desde la columna 'Contenido'
def extraer_fecha(texto):
    match = re.search(r"(\d{2})(\d{2})(\d{4})", str(texto))
    if match:
        return pd.to_datetime(f"{match.group(3)}-{match.group(2)}-{match.group(1)}")
    return pd.NaT

df["fecha"] = df["Contenido"].apply(extraer_fecha)
df = df.dropna(subset=["fecha"])

# 3. Cargar modelo FLAN-T5
## modelo FLAN-T5 localmente, descargado desde Hugging Face
modelo = "google/flan-t5-base"
tokenizer = T5Tokenizer.from_pretrained(modelo)
model = T5ForConditionalGeneration.from_pretrained(modelo)

# 4. Función de resumen
def resumir_con_flan(texto, instruccion="Haz un resumen claro y breve en español del siguiente contenido:"):
    if not texto or not isinstance(texto, str):
        return "Resumen no disponible."
    try:
        input_text = f"{instruccion} {texto}"
        inputs = tokenizer(input_text, return_tensors="pt", max_length=512, truncation=True)

        with torch.no_grad():
            outputs = model.generate(
                **inputs,
                max_length=60,
                num_beams=4,
                early_stopping=True
            )

        resumen = tokenizer.decode(outputs[0], skip_special_tokens=True)
        resumen = resumen.replace(instruccion, "").strip()
        return resumen
    except Exception as e:
        return f"⚠️ Error generando resumen: {e}"


# 5. Funciones de consulta
def ultimas_noticias_por_diario(df, n=3):
    noticias = (
        df.sort_values("fecha", ascending=False)
          .groupby("Diario")
          .head(n)
          .sort_values(["Diario", "fecha"], ascending=[True, False])
    )
    return noticias[["Diario", "Título", "fecha", "Enlace", "resumen_limpio"]]

def buscar_noticias_por_tema(df, tema, n=5):
    filtro = df["resumen_limpio"].str.contains(tema.lower(), case=False, na=False)
    resultados = df[filtro].sort_values("fecha", ascending=False).head(n)
    return resultados[["Diario", "Título", "fecha", "Enlace", "resumen_limpio"]]

# 6. Chatbot
def responder_pregunta(pregunta, df):
    pregunta = pregunta.lower()
    
    if "nuevo" in pregunta or "últimas" in pregunta:
        noticias = ultimas_noticias_por_diario(df)
        respuesta = "📰 Últimas noticias políticas por diario:\n\n"
        for _, row in noticias.iterrows():
            resumen_flan = resumir_con_flan(row["resumen_limpio"])
            respuesta += (
                f"- ({row['Diario']}) {row['Título']} ({row['fecha'].date()})\n"
                f"  📝 {resumen_flan}\n"
                f"  🔗 {row['Enlace']}\n\n"
            )
        return respuesta
    
    elif "dicen" in pregunta or "sobre" in pregunta:
        tema = pregunta.split("sobre")[-1].strip()
        noticias = buscar_noticias_por_tema(df, tema)
        if noticias.empty:
            return f"No encontré noticias recientes sobre “{tema}”."
        respuesta = f"🗞 Noticias recientes sobre “{tema}”:\n\n"
        for _, row in noticias.iterrows():
            resumen_flan = resumir_con_flan(row["resumen_limpio"])
            respuesta += (
                f"- ({row['Diario']}) {row['Título']} ({row['fecha'].date()})\n"
                f"  📝 {resumen_flan}\n"
                f"  🔗 {row['Enlace']}\n\n"
            )
        return respuesta
    
    else:
        return (
            "🤖 No entendí tu consulta. Puedes probar con:\n"
            "- ¿Qué hay de nuevo en política?\n"
            "- Dime las últimas noticias políticas\n"
            "- ¿Qué dicen los periódicos sobre Dina Boluarte?"
        )

# 7. Loop interactivo en consola
if __name__ == "__main__":
    print("🗣 Chatbot de noticias políticas con FLAN. Escribe 'salir' para terminar.\n")
    while True:
        user_input = input("Tú: ")
        if user_input.lower() in ["salir", "exit", "quit"]:
            print("👋 Hasta luego.")
            break
        respuesta = responder_pregunta(user_input, df)
        print("\n" + respuesta + "\n")


C:\Users\HP\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565


🗣 Chatbot de noticias políticas con FLAN. Escribe 'salir' para terminar.


🤖 No entendí tu consulta. Puedes probar con:
- ¿Qué hay de nuevo en política?
- Dime las últimas noticias políticas
- ¿Qué dicen los periódicos sobre Dina Boluarte?


🗞 Noticias recientes sobre “dina boluarte”:

- (Diario Gestión) Wallenberg Investments AB adelanta sus planes en Perú: Los sectores de interés (2025-07-11)
  📝 Haz un resumen claro y breve en espaol del siguiente contenido: Carlos rosales salas empresario sueco marcus wallenberg presidente de wallenberg investments ab que cuenta
  🔗 https://gestion.pe/economia/empresas/entrevista-a-marcus-wallenberg-presidente-del-directorio-de-saab-fabricante-del-gripen-sobre-su-visita-a-peru-noticia/

- (Diario El Comercio) CCL: bloqueo de vías en el sur pondría en riesgo S622 millones diarios y elevaría la inflación regional (2025-07-11)
  📝 Haz un resumen claro y breve en espaol del siguiente contenido: Update ec actualizado a instituto de economa y desarrollo 

Creando Dataset con resumenes reales

In [1]:
import pandas as pd

# Ruta del archivo original
ruta_entrada = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Inputs\noticias_politicas_procesadas.csv"
ruta_salida = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Outputs\noticias_benchmark_resumen.csv"

# Leer archivo original
df = pd.read_csv(ruta_entrada)
df.columns = df.columns.str.strip()

# Seleccionar 20 noticias con resumen_limpio no vacío
df_benchmark = df[["resumen_limpio"]].dropna().sample(20, random_state=42)

# Agregar columna vacía para que llenes los resúmenes a mano
df_benchmark["resumen_gold"] = ""

# Guardar archivo CSV para que lo edites manualmente
df_benchmark.to_csv(ruta_salida, index=False, encoding="utf-8-sig")

print(f"✅ Archivo generado en:\n{ruta_salida}")


✅ Archivo generado en:
C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Outputs\noticias_benchmark_resumen.csv


Leer tu archivo original y seleccionar 20 noticias

In [ ]:
import pandas as pd

# Cargar el archivo original
ruta = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Inputs\noticias_politicas_procesadas.csv"
df = pd.read_csv(ruta)
df.columns = df.columns.str.strip()

# Seleccionar 20 noticias aleatorias (o usa .head(20) si prefieres las primeras)
df_benchmark = df[["resumen_limpio"]].dropna().sample(20, random_state=42)

# Agregar columna vacía para que tú la llenes manualmente
df_benchmark["resumen_gold"] = ""

# Guardar como nuevo archivo para editarlo
df_benchmark.to_csv("noticias_benchmark_resumen.csv", index=False)
print("✅ Archivo generado: noticias_benchmark_resumen.csv")


comparar_mt5_vs_xlsum.py

In [10]:
import pandas as pd
import torch
from transformers import T5Tokenizer, T5ForConditionalGeneration
import evaluate

# === 1. Cargar benchmark con resumen_gold ===
import pandas as pd

# Ruta del archivo CSV con resumen_gold
ruta = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Outputs\noticias_benchmark_resumen.csv"

df = pd.read_csv(ruta, sep=";", engine="python", encoding="utf-8-sig")
df = df.dropna(subset=["resumen_limpio", "resumen_gold"])




# === 2. Definir modelos y prompts ===
modelos = {
    "mt5_base": {
        "ruta": "google/mt5-base",
        "prompt": "Resume en español:"
    },
    "mt5_xlsum": {
        "ruta": "csebuetnlp/mT5_multilingual_XLSum",
        "prompt": "resumir:"
    }
}

# === 3. Función para resumir con un modelo dado ===
def generar_resumen(modelo_id, tokenizer, model, texto, prompt):
    entrada = f"{prompt} {texto}"
    inputs = tokenizer(entrada, return_tensors="pt", truncation=True, max_length=512)
    with torch.no_grad():
        output = model.generate(inputs["input_ids"], max_length=60, num_beams=4)
    return tokenizer.decode(output[0], skip_special_tokens=True)

# === 4. Generar resúmenes y evaluar con ROUGE ===
rouge = evaluate.load("rouge")
resultados = {}

for nombre, config in modelos.items():
    print(f"🧠 Procesando modelo: {nombre}")
    tokenizer = T5Tokenizer.from_pretrained(config["ruta"])
    model = T5ForConditionalGeneration.from_pretrained(config["ruta"])

    df[f"resumen_{nombre}"] = df["resumen_limpio"].apply(
        lambda t: generar_resumen(nombre, tokenizer, model, t, config["prompt"])
    )

    # Calcular métricas ROUGE
    scores = rouge.compute(
        predictions=df[f"resumen_{nombre}"].tolist(),
        references=df["resumen_gold"].tolist()
    )
    resultados[nombre] = scores

# === 5. Mostrar resultados ===
for nombre, metricas in resultados.items():
    print(f"\n📊 ROUGE para {nombre}:")
    for k, v in metricas.items():
        print(f"{k}: {v:.4f}")


🧠 Procesando modelo: mt5_base


C:\Users\HP\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\HP\.cache\huggingface\hub\models--google--mt5-base. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
You are using the default legacy behav

🧠 Procesando modelo: mt5_xlsum


C:\Users\HP\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\HP\.cache\huggingface\hub\models--csebuetnlp--mT5_multilingual_XLSum. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
You are using a mode


📊 ROUGE para mt5_base:
rouge1: 0.0555
rouge2: 0.0091
rougeL: 0.0472
rougeLsum: 0.0494

📊 ROUGE para mt5_xlsum:
rouge1: 0.2424
rouge2: 0.0808
rougeL: 0.1610
rougeLsum: 0.1756


Interpretación rápida
1) mt5_xlsum tiene un rendimiento muy superior en todas las métricas, especialmente en ROUGE-1 y ROUGE-Lsum.
2) mt5_base genera resúmenes con muy bajo solapamiento léxico con los de referencia, indicando que:

2.1.) No está bien afinado para la tarea.
2.2.) O los resúmenes son muy genéricos.

In [8]:
!pip install rouge_score nltk


  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
   ---------------------------------------- 0.0/1.5 MB ? eta -:--:--
   -- ------------------------------------- 0.1/1.5 MB 6.4 MB/s eta 0:00:01
   ----- ---------------------------------- 0.2/1.5 MB 3.9 MB/s eta 0:00:01
   ----- ---------------------------------- 0.2/1.5 MB 3.9 MB/s eta 0:00:01
   ----- ---------------------------------- 0.2/1.5 MB 3.9 MB/s eta 0:00:01
   ----- ---------------------------------- 0.2/1.5 MB 3.9 MB/s eta 0:00:01
   ----- ---------------------------------- 0.2/1.5 MB 3.9 MB/s eta 0:00:01
   ----- ---------------------------------- 0.2/1.5 MB 3.9 MB/s eta 0:00:01
   ----- ---------------------------------- 0.2/1.5 MB 3.9 MB/s eta 0:00:01
   --------- ------------------------------ 0.4/1.5 MB 825.0 kB/s eta 0:00:02
   ------------ --------------------------- 0.5/1.5 MB 1.0 MB/s eta 0:00:01
   ------------ --------------------------- 0.5/1.5 MB 1.0 MB/s eta


[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\HP\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [11]:
pip install bert-score


     ---------------------------------------- 0.0/110.1 kB ? eta -:--:--
     --- ------------------------------------ 10.2/110.1 kB ? eta -:--:--
     --- ------------------------------------ 10.2/110.1 kB ? eta -:--:--
     --------------------------- --------- 81.9/110.1 kB 770.8 kB/s eta 0:00:01
     ------------------------------ ------ 92.2/110.1 kB 585.1 kB/s eta 0:00:01
     ------------------------------------ 110.1/110.1 kB 583.7 kB/s eta 0:00:00
   ---------------------------------------- 0.0/61.1 kB ? eta -:--:--
   ------ --------------------------------- 10.2/61.1 kB ? eta -:--:--
   --------------------------------- ------ 51.2/61.1 kB 871.5 kB/s eta 0:00:01
   ---------------------------------------- 61.1/61.1 kB 807.9 kB/s eta 0:00:00
   ---------------------------------------- 0.0/8.1 MB ? eta -:--:--
   ---------------------------------------- 0.0/8.1 MB ? eta -:--:--
   ---------------------------------------- 0.1/8.1 MB 812.7 kB/s eta 0:00:10
   -------------------


[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\HP\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Código para evaluación con BERTScore:

##BERTScore F1 mide la similaridad semántica entre el resumen generado y el de referencia.
l valor va de 0 a 1:
> 0.85: Excelente (usualmente solo alcanzado por humanos o modelos finamente ajustados).
0.75 - 0.85: Muy bueno.
0.65 - 0.75: Aceptable, especialmente si los textos son difíciles o hay variación en redacción.
< 0.65: Puede haber problemas de cobertura o comprensión semántica.

In [14]:
import evaluate

# Cargar métrica BERTScore
bertscore = evaluate.load("bertscore")

# Evaluar mt5_base
bert_mt5_base = bertscore.compute(predictions=df["resumen_mt5_base"].tolist(),
                                  references=df["resumen_gold"].tolist(),
                                  lang="es")

# Evaluar mt5_xlsum
bert_mt5_xlsum = bertscore.compute(predictions=df["resumen_mt5_xlsum"].tolist(),
                                   references=df["resumen_gold"].tolist(),
                                   lang="es")

# Mostrar resultados promedio
print("📊 BERTScore para mt5_base:")
print(f"Precision: {sum(bert_mt5_base['precision'])/len(bert_mt5_base['precision']):.4f}")
print(f"Recall:    {sum(bert_mt5_base['recall'])/len(bert_mt5_base['recall']):.4f}")
print(f"F1-score:  {sum(bert_mt5_base['f1'])/len(bert_mt5_base['f1']):.4f}")

print("\n📊 BERTScore para mt5_xlsum:")
print(f"Precision: {sum(bert_mt5_xlsum['precision'])/len(bert_mt5_xlsum['precision']):.4f}")
print(f"Recall:    {sum(bert_mt5_xlsum['recall'])/len(bert_mt5_xlsum['recall']):.4f}")
print(f"F1-score:  {sum(bert_mt5_xlsum['f1'])/len(bert_mt5_xlsum['f1']):.4f}")

#Apuntes:
#lang="es" indica que BERTScore usará un modelo multilingüe adecuado para español.
#La métrica más representativa es F1, que combina precisión y recall semántico.
#Esto te permitirá ver si aunque un modelo no use exactamente las mismas palabras (como FLAN), puede tener buena similaridad semántica.


Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


📊 BERTScore para mt5_base:
Precision: 0.5893
Recall:    0.5447
F1-score:  0.5655

📊 BERTScore para mt5_xlsum:
Precision: 0.7005
Recall:    0.6236
F1-score:  0.6589


Se usará mt5_xlsum como base para mejorar la los resumenes de noticias locales.
Siguientes pasos recomendados:
1. Ajustar los prompts para mejores resúmenes.
2. Evaluar en más ejemplos (aunque sea 50–100).
3. Crear al menos 200 ejemplos reales, para hacer fine-tuning en Google Colab con GPU gratuita.

In [15]:
pip install transformers pandas torch


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: C:\Users\HP\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [1]:
# 1. Librerías
import pandas as pd
import re
import torch
from transformers import T5Tokenizer, T5ForConditionalGeneration

# 2. Cargar CSV y preparar el DataFrame
ruta = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Inputs\noticias_politicas_procesadas.csv"
df = pd.read_csv(ruta)
df.columns = df.columns.str.strip()

# Verificación de columna 'resumen_limpio'
if 'resumen_limpio' not in df.columns:
    print("⚠️ 'resumen_limpio' no encontrado. Usando 'Contenido' como respaldo.")
    df['resumen_limpio'] = df['Contenido']

# Extraer fecha desde la columna 'Contenido'
def extraer_fecha(texto):
    match = re.search(r"(\d{2})(\d{2})(\d{4})", str(texto))
    if match:
        return pd.to_datetime(f"{match.group(3)}-{match.group(2)}-{match.group(1)}")
    return pd.NaT

df["fecha"] = df["Contenido"].apply(extraer_fecha)
df = df.dropna(subset=["fecha"])

# 3. Cargar modelo mT5_XLSum
modelo = "csebuetnlp/mT5_multilingual_XLSum"
tokenizer = T5Tokenizer.from_pretrained(modelo)
model = T5ForConditionalGeneration.from_pretrained(modelo)

# 4. Función de resumen con mT5_XLSum (prompt ajustado)
def resumir_con_mt5_xlsum(texto):
    if not texto or not isinstance(texto, str):
        return "Resumen no disponible."
    try:
        # Prompt mejorado con foco en hechos clave
        input_text = f"summarize: resume los hechos políticos más relevantes en el siguiente texto: {texto}"
        inputs = tokenizer(input_text, return_tensors="pt", max_length=512, truncation=True)

        with torch.no_grad():
            outputs = model.generate(
                **inputs,
                max_length=60,
                num_beams=4,
                early_stopping=True
            )

        resumen = tokenizer.decode(outputs[0], skip_special_tokens=True)
        return resumen.strip()
    except Exception as e:
        return f"⚠️ Error generando resumen: {e}"


# 5. Funciones de consulta
def ultimas_noticias_por_diario(df, n=3):
    noticias = (
        df.sort_values("fecha", ascending=False)
          .groupby("Diario")
          .head(n)
          .sort_values(["Diario", "fecha"], ascending=[True, False])
    )
    return noticias[["Diario", "Título", "fecha", "Enlace", "resumen_limpio"]]

def buscar_noticias_por_tema(df, tema, n=5):
    filtro = df["resumen_limpio"].str.contains(tema.lower(), case=False, na=False)
    resultados = df[filtro].sort_values("fecha", ascending=False).head(n)
    return resultados[["Diario", "Título", "fecha", "Enlace", "resumen_limpio"]]

# 6. Chatbot
def responder_pregunta(pregunta, df):
    pregunta = pregunta.lower()
    
    if "nuevo" in pregunta or "últimas" in pregunta:
        noticias = ultimas_noticias_por_diario(df)
        respuesta = "📰 Últimas noticias políticas por diario:\n\n"
        for _, row in noticias.iterrows():
            resumen = resumir_con_mt5_xlsum(row["resumen_limpio"])
            respuesta += (
                f"- ({row['Diario']}) {row['Título']} ({row['fecha'].date()})\n"
                f"  📝 {resumen}\n"
                f"  🔗 {row['Enlace']}\n\n"
            )
        return respuesta
    
    elif "dicen" in pregunta or "sobre" in pregunta:
        tema = pregunta.split("sobre")[-1].strip()
        noticias = buscar_noticias_por_tema(df, tema)
        if noticias.empty:
            return f"No encontré noticias recientes sobre “{tema}”."
        respuesta = f"🗞 Noticias recientes sobre “{tema}”:\n\n"
        for _, row in noticias.iterrows():
            resumen = resumir_con_mt5_xlsum(row["resumen_limpio"])
            respuesta += (
                f"- ({row['Diario']}) {row['Título']} ({row['fecha'].date()})\n"
                f"  📝 {resumen}\n"
                f"  🔗 {row['Enlace']}\n\n"
            )
        return respuesta
    
    else:
        return (
            "🤖 No entendí tu consulta. Puedes probar con:\n"
            "- ¿Qué hay de nuevo en política?\n"
            "- Dime las últimas noticias políticas\n"
            "- ¿Qué dicen los periódicos sobre Dina Boluarte?"
        )

# 7. Loop interactivo en consola
if __name__ == "__main__":
    print("🗣 Chatbot de noticias políticas con mT5_XLSum. Escribe 'salir' para terminar.\n")
    while True:
        user_input = input("Tú: ")
        if user_input.lower() in ["salir", "exit", "quit"]:
            print("👋 Hasta luego.")
            break
        respuesta = responder_pregunta(user_input, df)
        print("\n" + respuesta + "\n")


C:\Users\HP\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
You are using a model of type mt5 to instantiate a model of type t5. This is not supported for all configurations of models and can yield errors.


🗣 Chatbot de noticias políticas con mT5_XLSum. Escribe 'salir' para terminar.


🤖 No entendí tu consulta. Puedes probar con:
- ¿Qué hay de nuevo en política?
- Dime las últimas noticias políticas
- ¿Qué dicen los periódicos sobre Dina Boluarte?


📰 Últimas noticias políticas por diario:

- (Diario Correo) Un arancel de 50 al cobre hará caer exportación de Perú: Trump anunció que sería desde el 1 de agosto (2025-07-11)
  📝 En esta nueva serie, los periodistas de la BBC te ayudan a practicar una serie de reflexiones sobre los hechos políticos más importantes del año pasado.
  🔗 https://diariocorreo.pe/economia/un-arancel-de-50-al-cobre-hara-caer-exportacion-de-peru-trump-anuncio-que-seria-desde-el-1-de-agosto-noticia/

- (Diario Correo) Trabajadores deben recibir su gratificación de Fiestas Patrias hasta el 15 de julio (2025-07-11)
  📝 El martes de julio los trabajadores formales del sector privado esperan con expectativa un ingreso adicional que corresponde por ley a la fecha límite par

CÓDIGO PARA EVALUACIÓN COMPARATIVA

In [2]:
# Función para generar resumen con un prompt personalizado
def generar_resumen(texto, prompt="original"):
    if not texto or not isinstance(texto, str):
        return "Resumen no disponible."
    try:
        if prompt == "original":
            input_text = f"summarize: {texto}"
        elif prompt == "mejorado":
            input_text = f"summarize: resume los hechos políticos más relevantes en el siguiente texto: {texto}"
        else:
            return "❌ Prompt desconocido"

        inputs = tokenizer(input_text, return_tensors="pt", max_length=512, truncation=True)
        with torch.no_grad():
            outputs = model.generate(
                **inputs,
                max_length=60,
                num_beams=4,
                early_stopping=True
            )
        resumen = tokenizer.decode(outputs[0], skip_special_tokens=True)
        return resumen.strip()
    except Exception as e:
        return f"⚠️ Error generando resumen: {e}"


EVALUACIÓN VISUAL

In [3]:
# Seleccionamos 5 ejemplos aleatorios del dataset
ejemplos = df.sample(5, random_state=42)[["Título", "resumen_limpio"]]

# Mostrar resumen con ambos prompts
for i, row in ejemplos.iterrows():
    print(f"📰 TÍTULO: {row['Título']}\n")
    print(f"📄 Texto original:\n{row['resumen_limpio'][:500]}...\n")
    print(f"✏️ Prompt original:\n{generar_resumen(row['resumen_limpio'], prompt='original')}\n")
    print(f"✅ Prompt mejorado:\n{generar_resumen(row['resumen_limpio'], prompt='mejorado')}\n")
    print("="*80)


📰 TÍTULO: Qué se discutió en la reunión entre Dina Boluarte y Julio Velarde sobre la economía peruana?

📄 Texto original:
Boluarte zegarra sostuvo una reunión el jueves con el jefe del banco central de reserva del perú bcrp julio velarde flores en palacio de gobierno el objetivo del encuentro fue el seguir articulando acciones orientadas a fortalecer el crecimiento de la economía peruana informó la presidencia de la república la estrecha coordinación entre el poder ejecutivo y el ente emisor resulta fundamental para asegurar la sostenibilidad del desempeño económico del perú dado que ambas instituciones son respons...

✏️ Prompt original:
El jefe del banco central de reserva del perú bcrp julio velarde flores foto presidencia recibió este jueves en palacio de gobierno a la jefa de estado y titular de el Banco Central de Reserva del Perú, ed

✅ Prompt mejorado:
El jefe del banco central de reserva del perú bcrp julio velarde flores foto presidencia recibió este jueves en palacio de gobi

Script con ROUGE y BERTScore

In [5]:
import pandas as pd
import torch
from transformers import T5Tokenizer, T5ForConditionalGeneration
import evaluate

# === 1. Cargar benchmark ===
ruta = r"C:\Users\HP\OneDrive - Grupo Alicorp\Escritorio\uni\Curso PLN\Cursos\PLN - ExamenFinal\Outputs\noticias_benchmark_resumen.csv"
df = pd.read_csv(ruta, sep=";", engine="python", encoding="utf-8-sig")
df = df.dropna(subset=["resumen_limpio", "resumen_gold"])

# === 2. Definir modelos y prompts ===
modelos = {
    "mt5_base": {
        "ruta": "google/mt5-base",
        "prompt": "Resume en español:"
    },
    "mt5_xlsum": {
        "ruta": "csebuetnlp/mT5_multilingual_XLSum",
        "prompt": "resumir:"
    },
    "mt5_xlsum_prompt": {
        "ruta": "csebuetnlp/mT5_multilingual_XLSum",
        "prompt": "Por favor, resume en español de forma clara y concisa el contenido de esta noticia:"
    }
}

# === 3. Función de resumen ===
def generar_resumen(modelo_id, tokenizer, model, texto, prompt):
    entrada = f"{prompt} {texto}"
    inputs = tokenizer(entrada, return_tensors="pt", truncation=True, max_length=512)
    with torch.no_grad():
        output = model.generate(inputs["input_ids"], max_length=60, num_beams=4)
    return tokenizer.decode(output[0], skip_special_tokens=True)

# === 4. Métricas ===
rouge = evaluate.load("rouge")
bertscore = evaluate.load("bertscore")
resultados = {}

# === 5. Evaluar modelos ===
for nombre, config in modelos.items():
    print(f"\n🧠 Procesando modelo: {nombre}")
    tokenizer = T5Tokenizer.from_pretrained(config["ruta"])
    model = T5ForConditionalGeneration.from_pretrained(config["ruta"])

    df[f"resumen_{nombre}"] = df["resumen_limpio"].apply(
        lambda t: generar_resumen(nombre, tokenizer, model, t, config["prompt"])
    )

    # ROUGE
    rouge_scores = rouge.compute(
        predictions=df[f"resumen_{nombre}"].tolist(),
        references=df["resumen_gold"].tolist()
    )

    # BERTScore
    bert_scores = bertscore.compute(
        predictions=df[f"resumen_{nombre}"].tolist(),
        references=df["resumen_gold"].tolist(),
        lang="es"
    )

    resultados[nombre] = {
        "ROUGE": rouge_scores,
        "BERTScore": {
            "precision": sum(bert_scores["precision"]) / len(bert_scores["precision"]),
            "recall": sum(bert_scores["recall"]) / len(bert_scores["recall"]),
            "f1": sum(bert_scores["f1"]) / len(bert_scores["f1"])
        }
    }

# === 6. Mostrar resultados ===
for nombre, res in resultados.items():
    print(f"\n📊 Resultados para {nombre}:")
    print("🔹 ROUGE:")
    for k, v in res["ROUGE"].items():
        print(f"  {k}: {v:.4f}")
    print("🔹 BERTScore:")
    for k, v in res["BERTScore"].items():
        print(f"  {k}: {v:.4f}")



🧠 Procesando modelo: mt5_base


You are using a model of type mt5 to instantiate a model of type t5. This is not supported for all configurations of models and can yield errors.



🧠 Procesando modelo: mt5_xlsum


You are using a model of type mt5 to instantiate a model of type t5. This is not supported for all configurations of models and can yield errors.



🧠 Procesando modelo: mt5_xlsum_prompt


You are using a model of type mt5 to instantiate a model of type t5. This is not supported for all configurations of models and can yield errors.



📊 Resultados para mt5_base:
🔹 ROUGE:
  rouge1: 0.0553
  rouge2: 0.0092
  rougeL: 0.0474
  rougeLsum: 0.0495
🔹 BERTScore:
  precision: 0.5893
  recall: 0.5447
  f1: 0.5655

📊 Resultados para mt5_xlsum:
🔹 ROUGE:
  rouge1: 0.2432
  rouge2: 0.0808
  rougeL: 0.1607
  rougeLsum: 0.1745
🔹 BERTScore:
  precision: 0.7005
  recall: 0.6236
  f1: 0.6589

📊 Resultados para mt5_xlsum_prompt:
🔹 ROUGE:
  rouge1: 0.2276
  rouge2: 0.0753
  rougeL: 0.1581
  rougeLsum: 0.1667
🔹 BERTScore:
  precision: 0.6986
  recall: 0.6254
  f1: 0.6590


### Interpretacion
1. El modelo mt5_xlsum sin modificar sigue siendo el mejor en ROUGE, sin embargo;
2. Con el prompt mejorado (mt5_xlsum_prompt) igualas (o mejoras muy ligeramente) en BERTScore, que capta mejor la calidad semántica.

| Modelo             | ROUGE-Lsum | BERTScore F1 | Comentario                                                     |
| ------------------ | ---------- | ------------ | -------------------------------------------------------------- |
| `mt5_base`         | 0.0495     | 0.5655       | Bajo rendimiento general. No entrenado para resumen.           |
| `mt5_xlsum`        | **0.1745** | 0.6589       | El mejor en ROUGE (más cercanía estructural).                  |
| `mt5_xlsum_prompt` | 0.1667     | **0.6590**   | Ligeramente mejor en BERTScore, gracias al prompt más natural. |


### En resumen:

- **ROUGE** mide solapamiento de *n-gramas*, útil para evaluar **fidelidad literal** al texto de referencia.
- **BERTScore** usa *embeddings* y es más tolerante a **reescrituras semánticas**, premiando la **calidad del lenguaje**.

**Entonces:**

> Si se buscan resúmenes más **legibles y naturales**, el prompt mejorado (`mt5_xlsum_prompt`) es una **gran elección**.


#### Pasos para desplegar el modelo mt5_xlsum_prompt en una interfaz local

In [6]:
#guardar el modelo y tokenizer
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

modelo = "csebuetnlp/mT5_multilingual_XLSum"
tokenizer = AutoTokenizer.from_pretrained(modelo)
model = AutoModelForSeq2SeqLM.from_pretrained(modelo)

tokenizer.save_pretrained("mt5_xlsum_prompt_model")
model.save_pretrained("mt5_xlsum_prompt_model")


C:\Users\HP\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\transformers\convert_slow_tokenizer.py:564: UserWarning: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.
  warnings.warn(
C:\Users\HP\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\transformers\modeling_utils.py:3685: UserWarning: Moving the following attributes in the config to the generation config: {'max_length': 84, 'num_beams': 4, 'length_penalty': 0.6, 'no_repeat_ngram_size': 2}. You are seeing this warning because you've set generation para